enzy_htp/analysis/spi.py

```diff
@@ -0,0 +1,115 @@
+
+
+@dispatch
+def spi_metric(
+    stru : StructureEnsemble,
+    ligand : Ligand,
+    pocket_sele: str 
+) -> List[float]:
+    """Calculates the spi metric for a StructureEnsemble using a pymol-formatted pocket sele str. 
+    Note: The pocket_sele expression is applied to each Structure(), so it is recommended to use a selection which specifies
+    individual Residue()'s as the expression will be re-evaluated for each Structure() geometry.
+
+    Args:
+        stru: The StructureEnsemble() to analyze.
+        ligand: The Ligand() which will be the numerator in the spi metric.
+        pocket_sele: A pymol-formatted sele str which defines the denominator in the spi metric.
+
+    Returns:
+        A List[float] of len(StructureEnsemble.structures), with each value being a different spi metric for the nth Structure(). 
+    """
+    
+    if isinstance(stru, Structure):
+        #TODO(CJ): put some kind of thing here
+        stru = stru.from_single_stru( stru )
+
+    result = list()
+    for ss in stru.structures:
+        result.append(
+            spi_metric( ss, ligand, pocket_sele )
+        )
+
+    return result
+
+@dispatch
+def spi_metric(
+    stru : StructureEnsemble,
+    ligand : Ligand,
+    pocket: List[Residue],
+) -> List[float]:
+    """Calculates the spi metric for a StructureEnsemble using a List[Residue] that define the denominator of the 
+    metric. This is the PREFERRED method for calculating spi.
+
+    Args:
+        stru: The StructureEnsemble() to analyze.
+        ligand: The Ligand() which will be the numerator in the spi metric.
+        pocket: A List[Residue] which serves as the pocket or SASA denominator of the spi metric.
+
+    Returns:
+        A List[float] of len(StructureEnsemble.structures), with each value being a different spi metric for the nth Structure(). 
+    """
+    pocket_sele = " or ".join(map(
+        lambda rr: f"( chain {rr.parent.name} and resi {rr.idx} )",
+        pocket
+    ))
+
+    result = list()
+    for ss in stru.structures:
+        result.append(
+            spi_metric( ss, ligand, pocket_sele )
+        )
+
+    return result
+
+
+
+@dispatch
+def spi_metric( stru: Structure, 
+            ligand: Ligand,
+            pocket_sele:str ) -> float:
+    """Calculates the spi metric for a single Structure using a pymol-formatted pocket sele str. 
+    Note: The pocket_sele expression is applied to the Structure(), so it is recommended to use a selection which specifies
+    individual Residue()'s.
+
+    Args:
+        stru: The Structure() to analyze.
+        ligand: The Ligand() which will be the numerator in the spi metric.
+        pocket_sele: A pymol-formatted sele str which defines the denominator in the spi metric.
+
+    Returns:
+        A List[float] of len(StructureEnsemble.structures), with each value being a different spi metric for the nth Structure(). 
+    """
+
+    (lig_chain, lig_idx) = ligand.key()
+    session = interface.pymol.new_session()
```

Comment:
Please use the context manager for pymol session here

Reply:
got it!

Reply:
Added context manager